internal/restapi/providers/v1/handler.go

```diff
@@ -0,0 +1,262 @@
+	Versions	[]VersionItem	`json:"versions"`
+}
+
+// Structs to load response into (for a provider's metadata)
+
+type PlatformMetadataResponse struct {
+	Protocols     	Protocols 	`json:"protocols"`
+	OS            	string    	`json:"os"`
+	Arch          	string    	`json:"arch"`
+	Filename      	string    	`json:"filename"`
+	DownloadURL   	string    	`json:"download_url"`
+	ShasumsURL    	string    	`json:"shasums_url"`
+	ShasumsSigURL	string    	`json:"shasums_signature_url"`
+	Shasum        	string    	`json:"shasum"`
+	SigningKeys   	SigningKeys	`json:"signing_keys"`
+}
+
+
+func New() *providersV1HttpService {
+	return &providersV1HttpService{}
+}
+
+func (h *providersV1HttpService) GetHttpHandler(mountPath string) http.Handler {
+	router := h.createRouter(mountPath)
+	return handlers.CombinedLoggingHandler(os.Stdout, router)
+}
+
+func (h *providersV1HttpService) createRouter(mountPath string) *mux.Router {
+	prefix := fmt.Sprintf("%s/v1", mountPath)
+	log.Printf("prefix for registry implementation: %s", prefix)
+	r := mux.NewRouter()
+	r.Handle("/healthz", h.healthHandler()).Methods(http.MethodGet)
+	sr := r.PathPrefix(prefix).Subrouter()
+	sr.Use(otelmux.Middleware("providers-v1"))
+	sr.StrictSlash(true)
+	sr.Handle("/{organization_name}/{name}/versions", h.getProviderVersionHandler()).Methods(http.MethodGet)
+	sr.Handle("/{organization_name}/{name}/{version}/download/{os}/{arch}", h.downloadProviderHandler()).Methods(http.MethodGet)
+	return r
+}
+
+func (h *providersV1HttpService) healthHandler() http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		h.responseHandler.Write(rw, "OK", http.StatusOK)
+	})
+}
+
+func loadJSONData() map[string]ProviderData {
+	filePath := "./random.json"
+	data, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		panic(err)
+	}
+	var obj map[string]ProviderData
+
+	if err := json.Unmarshal(data, &obj); err != nil {
+		panic(err)
+	}
+
+	return obj
+}
+
+// GetProviderVersionHandler will return a list of available versions for a given provider.
+// This signifies to the requesting CLI if that provider is available to consume from the registry.
+// Will return a 404 if a non-existent organization and/or provider is requested.
+// This handler complies with the following implementation from the provider protocol
+// https://developer.hashicorp.com/terraform/internals/provider-registry-protocol#list-available-versions
+func (h *providersV1HttpService) getProviderVersionHandler() http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		log.Printf("getProviderVersionHandler")
+		
+		providerName := GetProviderNameFromRequest(r)
+
+		ctx := r.Context()
+		span := trace.SpanFromContext(ctx)
+		span.AddEvent("Returning a list of available versions for a provider", trace.WithAttributes(attribute.String("Provider Name", providerName)))
+		span.SetAttributes(
+			attribute.String("provider.name", providerName),
+		)
+
+		obj := loadJSONData()
```

Comment:
I'd suggest introducing a new package `internal/providers/services` to handle the retrieval of information from whatever source about the providers

The restAPI then just uses those API to format the response in the correct format. In this way you can then switch out the JSON data access client for a DynamoDB access client without changing this code.

Take a look at how we've done this for modules.

Reply:
Introduced a new package `internal/provider/services` for handling the data retrieval aspects.